docs/design/core/tagged-unions/README.md

```diff
@@ -0,0 +1,150 @@
+class AttributeValue {
+    private String s;
+    private String n;
+    private SdkBytes b;
+    private List<String> ss;
+    private List<String> ns;
+    private List<SdkBytes> bs;
+    
+    // ...
+}
+```
+
+Currently, customers are left to perform ad-hoc matching on the value to first determine what kind of value is provided for a union, and then dispatch to branching code to interact with the value. Rather than ask customers to do this themselves, we should provide abstractions that make it easy for customers to use these values.
+
+## Goal
+
+The goal of this project is to provide a backwards-compatible customer-friendly abstraction for existing and new union types within the 2.x SDK.
+
+## Prior Art
+
+We currently implement friendly tagged unions in at least two external places within the SDK already:
+1. The `software.amazon.awssdk.core.document.Document` type.
+2. Event stream types, like TranscribeStreaming's `software.amazon.awssdk.services.transcribestreaming.model.AudioStream` type.
+
+And at least one internal place:
+1. The `software.amazon.awssdk.protocols.jsoncore.JsonNode` type.
+
+## Union Type Definition
+
+We will follow the patterns set by `Document` and `JsonNode`, because the pattern used by `AudioStream` would not be backwards-compatible with existing types and does not function particularly well for union type members that are not structures.
+
+Existing structures primarily contain the following methods:
+1. A static `builder()` method for creating the structure and setting its mutually exclusive members.
+2. "Get" methods for each member in the union.
+3. "Has" methods for each collection or map member, to allow for differentiating between absent and missing values.
+
+In addition to the above methods, new and existing union types will have the following additional methods:
+1. A static `T fromN(...)` method for creating the structure with its single exclusive member initialized. e.g. `AttributeValue s = AttributeValue.fromS("string")`
+2. A `<T> T accept(X.Visitor<T>)` method for transforming the union type. e.g. `String transformedOutput = getItemResponse.item().accept(new AttributeValue.Visitor<String>() {...})`
+3. A `void accept(X.VoidVisitor)` method for branching out on the union type. e.g. `getItemResponse.item().accept(new AttributeValue.VoidVisitor() { ... })`
+4. A `X.Type type()` method for determining the type contained within the union. e.g. `AttributeValue.Type type = getItemResponse.item().type()`
+
+For example, an `AttributeValue` union type containing `String s`, `String n` and `SdkBytes b` would have the following public methods:
+```java
+class AttributeValue {
+    static AttributeValue.Builder builder();
+    static fromS(String s);
+    static fromN(String s);
+    static fromB(String s);
+    
+    String s();
+    String n();
+    SdkBytes b();
+
+    AttributeValue.Type type();
+    void accept(AttributeValue.VoidVisitor visitor);
+    <T> T accept(AttributeValue.Visitor<T> visitor);
```

Comment:
I was wondering, should we have these? The customer can do a switch based on the `type()` for something equivalent.

Reply:
Conclusion: They might be nice, but the void visitor is basically just a switch on the type() and the visitor<T> is just a switch expression on the type() (added in Java 17). Admittedly, not a lot of customers are on 17 right now, but it would be good not to add it if we can avoid it. We'll end up waiting for customer feedback on this one before we add it.